src/censored.jl

```diff
@@ -0,0 +1,278 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The density function (or mass function for discrete `d`) of the censored distribution is
+```math
+f(x; d, l, u) = \\begin{cases}
+    P_d(x \\le l), & x \\le l \\\\
+    P_d(x),         & l < x < u \\\\
+    P_d(x \\ge u), & x \\ge u \\\\
+  \\end{cases},
+```
+where ``P_d(x)`` is the density (or mass) function of `d`, and ``P_d(x \\le b)`` is the
+cumulative distribution function of ``d`` evaluated at ``x = b``.
+If ``x`` is a random variable from ``d``, then `clamp(x, l, u)` is a random variable from
+its censored version. Note that this implies that even if ``d`` is continuous, its censored
+form has discrete mixture components at the bounds.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real} = Censored(d, l, u)
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        l ≤ u || error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+Censored{D,S,T}(l::T, u::T, params...) where {D,S,T} = Censored(D(params...), l, u)
```

Comment:
Is there a strong argument for adding this constructor? Since users should not call `Censored` and it is not exported (IMO correctly), we don't have to provide any convenience constructors. Moreover, often you might want to specify keyword arguments such as `check_args` as well.
```suggestion
```

Reply:
As noted in https://github.com/JuliaStats/Distributions.jl/pull/1470#discussion_r782947768, the `params` docstring requires:
```julia
  params(d::UnivariateDistribution)

  Return a tuple of parameters. Let d be a distribution of type D, then
  D(params(d)...) will construct exactly the same distribution as d.
```

Reply:
It doesn't work in general though - it definitely doesn't work for `Truncated` and I would assume there are other distributions as well where it is not satisfied. I still tend to think that this can't be expected for such unexported fallback definitions. Unfortunately the docstring is not completely clear but intuitively I would have assumed that `D` also does not include any type parameters. Probably also since (outer) constructors with type parameters are very very rare in Distributions.